x/reporter/reporter.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reporter
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"math/rand"
+	"net/http"
+)
+
+var debugLog log.Logger = *log.New(io.Discard, "", 0)
+var httpClient = &http.Client{}
+
+type Record struct {
+	collectorURL      string
+	successSampleRate float64
+	failureSampleRate float64
+	success           bool
+	record            interface{}
+}
```

Comment:
We need to separate what is a report and what's a reporter.

We will need to specify the exact format of a report, which can be a struct, but for now it's any, but you need a success field:
```go
type Report struct {
  any
  success bool
}
```

The reporter must be an interface, so that anyone can implement one:
```go
type Reporter struct {
    Report(r Report) error
}
```
Perhaps we need to rename the method to something else, like `Register`. I'm hesitant of using `Send` because some reporters may batch them and send later.

Then provide an implementation that sends to a remote collector. It should have the url, and rates in the implementation structure, so one can call `reporter.Report(report)`.

Reply:
Just wanted to clarify something; Right now I am separating URL and sample rates from the log record / JSON that include the connectivity test error log, and having that data in a higher level structure. From what I understand here, you'd prefer to have that information in the JSON report? such as in this example:
```
	jsonData := `{
		"meta": {
			"successSampleRate": 1.0,
			"failureSampleRate": 0.0
			"collectorURL": "https://example-collector.com/collect"
		},
		"proxy": "192.168.1.1:65000",
		"transport": "shadowsocks",
		"resolver": "8.8.8.8:53",
		"proto": "tcp",
		"prefix": "HTTP1/1",
		"time": "2021-01-01T00:00:00Z",
		"durationMs": 100,
		"error": {
			"operation": "read",
			"posixError": "ETIMEDOUT",
			"msg": "i/o timeout"
		}
	}`  
```

Reply:
There's a lot of work to be done there. It's probably better to discuss in a doc.

For now it can be something like this:

```json
{
		"sampling_fraction": 1.0,
		"proxy": "192.168.1.1:65000",
		"transport": "shadowsocks",
		"resolver": "8.8.8.8:53",
		"proto": "tcp",
		"prefix": "HTTP/1.1",
		"time": "2021-01-01T00:00:00Z",
		"durationMs": 100,
		"error": {
			"operation": "read",
			"posixError": "ETIMEDOUT",
			"msg": "i/o timeout"
		}
}
```

- No need to specify the collector URL.
- You only need to specify the sampling fraction for this specific request.

Later we will need to extract what is test-specific (resolver) and transport-specific (shadowsocks, prefix).
For the DNS test, we should include the test domain. Not that relevant for an encrypted transport, but it will help in other cases.

Reply:
been working on a design doc for this and have updated the code to reflect some of the proposed design. I will share that shortly with you.